src/server/listener.cc

```diff
@@ -340,5 +389,82 @@ Listener::dispatchPeer(const std::shared_ptr<Peer>& peer) {
 
 }
 
+#ifdef PISTACHE_USE_SSL
+
+static SSL_CTX *ssl_create_context(std::string cert, std::string key, bool use_compression)
+{
+    const SSL_METHOD    *method;
+    SSL_CTX             *ctx;
+
+    method = SSLv23_server_method();
+
+    ctx = SSL_CTX_new(method);
+    if (ctx == NULL)
+    {
+        ERR_print_errors_fp(stderr);
+        throw std::runtime_error("Cannot setup SSL context");
+    }
+
+    if (!use_compression)
+    {
+        /* Disable compression to prevent BREACH and CRIME vulnerabilities. */
+        if (!SSL_CTX_set_options(ctx, SSL_OP_NO_COMPRESSION)) {
+            ERR_print_errors_fp(stderr);
+            throw std::runtime_error("Cannot disable compression");
+        }
+    }
+
+    SSL_CTX_set_ecdh_auto(ctx, 1);
+
+    if (SSL_CTX_use_certificate_file(ctx, cert.c_str(), SSL_FILETYPE_PEM) <= 0) {
+        ERR_print_errors_fp(stderr);
+        throw std::runtime_error("Cannot load SSL certificate");
+    }
+
+    if (SSL_CTX_use_PrivateKey_file(ctx, key.c_str(), SSL_FILETYPE_PEM) <= 0) {
+        ERR_print_errors_fp(stderr);
+        throw std::runtime_error("Cannot load SSL private key");
+    }
+
+    if (!SSL_CTX_check_private_key(ctx)) {
+        ERR_print_errors_fp(stderr);
+        throw std::runtime_error("Private key does not match public key in the certificate");
+    }
+
+    return ctx;
+}
+
+void
+Listener::setupSSLAuth(std::string ca_file, std::string ca_path)
+{
+    const char *__ca_file = NULL;
+    const char *__ca_path = NULL;
+
+    if (!ca_file.empty())
+        __ca_file = ca_file.c_str();
+    if (!ca_path.empty())
+        __ca_path = ca_path.c_str();
+
```

Comment:
It should check that `Listener::setupSSL` has been already called, if it hasn't the program crashes.